tests/MongoDB.Driver.Core.Tests/Core/Authentication/AvsSignatureVersion4Tests.cs

```diff
@@ -0,0 +1,130 @@
+            requestHeaders["X-Amz-Date"] = timestamp;
+            requestHeaders["Content-Length"] = "42";
+            requestHeaders["Host"] = "iam.testhost.com";
+            requestHeaders["X-MongoDB-GS2-CB-Flag"] = "n";
+            requestHeaders["X-MongoDB-Server-Nonce"] = "123";
+            var expected = "content-length;content-type;host;x-amz-date;x-amz-security-token;x-mongodb-gs2-cb-flag;x-mongodb-server-nonce";
+
+            var actual = AwsSignatureVersion4.GetSignedHeaders(requestHeaders);
+
+            actual.Should().Be(expected);
+        }
+
+        [Fact]
+        public void SignRequestShouldWorkAsExpected()
+        {
+            var date = new DateTime(2020, 03, 12, 14, 23, 46);
+            var accessKey = "permanentuser";
+            var secretKey = "FAKEFAKEFAKEFAKEFAKEfakefakefakefakefake";
+            var salt = new byte[] { 64, 230, 20, 164, 223, 96, 92, 144, 3, 240, 27, 110, 97, 65, 200, 11, 157, 162, 141, 4, 149, 86, 91, 108, 189, 194, 100, 90, 249, 219, 155, 235, };
+            var host = "sts.amazonaws.com";
+            var expected = Tuple.Create(
+                "AWS4-HMAC-SHA256 " +
+                "Credential=permanentuser/20200312/us-east-1/sts/aws4_request, " +
+                "SignedHeaders=content-length;content-type;host;x-amz-date;x-mongodb-gs2-cb-flag;x-mongodb-server-nonce, " +
+                "Signature=6872b9199b47dc983a95f9113a096c9b4e63bb6ddf39030161b1f092ab616df2",
+                "20200312T142346Z");
+
+            var actual = AwsSignatureVersion4.SignRequest(date, accessKey, secretKey, securityToken: null, salt, host);
+
+            actual.Should().Be(expected);
+        }
+
+        [Fact]
+        public void SignRequestWithSessionTokenShouldWorkAsExpected()
+        {
+            var date = new DateTime(2020, 03, 12, 14, 23, 46);
+            var accessKey = "permanentuser";
+            var secretKey = "FAKEFAKEFAKEFAKEFAKEfakefakefakefakefake";
+            var sessionToken = "MXUpbuzwzPo67WKCNYtdBq47taFtIpt+SVx58hNx1/jSz37h9d67dtUOg0ejKrv83u8ai+VFZxMx=";
+            var salt = new byte[] { 64, 230, 20, 164, 223, 96, 92, 144, 3, 240, 27, 110, 97, 65, 200, 11, 157, 162, 141, 4, 149, 86, 91, 108, 189, 194, 100, 90, 249, 219, 155, 235, };
+            var host = "sts.amazonaws.com";
+            var expected = Tuple.Create(
+                "AWS4-HMAC-SHA256 " +
+                "Credential=permanentuser/20200312/us-east-1/sts/aws4_request, " +
+                "SignedHeaders=content-length;content-type;host;x-amz-date;x-amz-security-token;x-mongodb-gs2-cb-flag;x-mongodb-server-nonce, " +
+                "Signature=d60ee7fe01c82631583a7534fe017e1840fd5975faf1593252e91c54573a93ae",
+                "20200312T142346Z");
+
+            var actual = AwsSignatureVersion4.SignRequest(date, accessKey, secretKey, sessionToken, salt, host);
+
+            actual.Should().Be(expected);
+        }
+    }
+}
```

Comment:
For the unit test in Java, I used the values in [AWS's documentation](https://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html) as opposed to the values returned from the mock servers. This was to ensure that my code conformed to AWS's signature v4 process rather than the mocks rendition of it, should they differ. It might be worthwhile to the same as a cautionary measure.